sdk/src/Core/Amazon.Runtime/Internal/Util/StreamingUtf8JsonReader.cs

```diff
@@ -0,0 +1,189 @@
+﻿/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ * 
+ *  http://aws.amazon.com/apache2.0
+ * 
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Text.Json;
+using System.Threading.Tasks;
+
+namespace Amazon.Runtime.Internal.Util
+{
+    /// <summary>
+    /// A wrapper around Utf8JsonReader that can handle reading from a stream.
+    /// </summary>
+    public ref struct StreamingUtf8JsonReader
+    {
+        private Utf8JsonReader _reader;
+        private JsonReaderOptions _options;
+        /// <summary>
+        /// The UTF8JsonReader attached to the instance. 
+        /// </summary>
+        public Utf8JsonReader Reader
+        {
+            get
+            {
+                return _reader;
+            }
+        }
+
+        private Stream _stream;
+        private byte[] _buffer;
+
+        public StreamingUtf8JsonReader(Stream stream)
+        {
+            if (stream is null)
+                throw new ArgumentException("Stream must not be null. Please initialize a stream and pass it into the constructor.");
+
+            _stream = stream;
+            _buffer = ArrayPool<byte>.Shared.Rent(AWSConfigs.StreamingUtf8JsonReaderBufferSize ?? 4096);
+            _options = new JsonReaderOptions
+            {
+                AllowTrailingCommas = true
+            };
+            // need to initialize the reader even if the buffer is empty because auto-default of unassigned fields is only 
+            // supported in C# 11+
+            _reader = new Utf8JsonReader(_buffer, _options);
+            SkipUtf8Bom(ref _buffer);
+        }
+
+        private void SkipUtf8Bom(ref byte[] buffer)
```

Comment:
Realizing this method does more then just Skip the Ut8Bom it is also what is in charge of providing the initial content for the buffer I think you should just collapse this method with the constructor. That also gets rid of the awkward `_reader` creating in the constructor that gets recreated here.  Something like this:
```
public StreamingUtf8JsonReader(Stream stream)
{
    if (stream is null)
        throw new ArgumentException("Stream must not be null. Please initialize a stream and pass it into the constructor.");

    _stream = stream;
    _buffer = ArrayPool<byte>.Shared.Rent(4096);
    _options = new JsonReaderOptions
    {
        AllowTrailingCommas = true
    };

    // Read the initial buffer from the stream and skip over UTF8 BOM if it exists.
    int bytesRead = _stream.Read(_buffer, 0, _buffer.Length);
    int start = 0;
    if (_buffer.AsSpan().StartsWith(JsonConstants.Utf8Bom))
    {
        start += JsonConstants.Utf8Bom.Length;
        bytesRead -= Utf8Bom.Length;
    }

    _reader = new Utf8JsonReader(_buffer.AsSpan(start, bytesRead), isFinalBlock: bytesRead == 0, new JsonReaderState(_options));
}
```

Reply:
that's a good point about the awkward `_reader` recreation. I will  just collapse it at this point.